addons/package_hierarchy/models/stock_quant_package.py

```diff
@@ -39,12 +39,47 @@ def _check_package_recursion(self):
 
     def _check_not_multi_location(self):
         if not self.env.context.get("bypass_multi_location_check", False):
-            for package in self:
-                locations = package.children_quant_ids.mapped('location_id')
-                if len(locations) > 1:
-                    raise ValidationError(_('Package cannot be in multiple '
-                                            'locations:\n%s\n%s') % (package.name,
-                                                                    ', '.join( [l.name for l in locations])))
+            packages = self.filtered(lambda p: p.package_id)
+            parent_packages = self.filtered(lambda p: not p.package_id)
+            multi_package_locations = False
+            multi_parent_package_locations = False
+            total_packages = []
+            if parent_packages:
+                parent_package_query = """
+                    SELECT COUNT(Distinct sq.location_id),pck.package_id
+                    FROM stock_quant sq
+                    JOIN stock_quant_package pck
+                    ON sq.package_id = pck.id
+                    WHERE pck.package_id in %(package_ids)s
+                    GROUP BY pck.package_id
+                    HAVING COUNT(Distinct sq.location_id) > 1;
+                """
+                self.env.cr.execute(
+                    parent_package_query, {"package_ids": tuple(parent_packages.ids)},
+                )
+                multi_parent_package_locations = self.env.cr.fetchall()
+            if packages:
+                package_query = """
+                    SELECT COUNT(Distinct sq.location_id),sq.package_id
+                    FROM stock_quant sq
+                    WHERE sq.package_id in %(package_ids)s
+                    GROUP BY sq.package_id
+                    HAVING COUNT(Distinct sq.location_id) > 1;
+                """
+                self.env.cr.execute(
+                    package_query, {"package_ids": tuple(packages.ids)},
+                )
+                multi_package_locations = self.env.cr.fetchall()
+            if multi_parent_package_locations:
+                total_packages += [result[1] for result in multi_parent_package_locations]
+            if multi_package_locations:
+                total_packages += [result[1] for result in multi_package_locations]
+            if total_packages:
+                all_multi_location_packages = self.browse(total_packages)
+                raise ValidationError(
+                    _("Package cannot be in multiple locations:\n%s") %
+                    (", ".join([p.name for p in all_multi_location_packages]))
```

Comment:
The error message no longer shows the locations where the packages are in.

Might be worth listing each package, and then all the locations it is in, so the issue is easier to resolve.

Reply:
Actually nevermind, as we discussed the user can now go to the package to find this info. But can we reword the error a bit. E.g.:

```
The following packages cannot be set into multiple locations:

xxx
```